### PR TITLE
Add begin-of-run interface to initialize states and actions

### PR DESCRIPTION
**Pull request by sethrj**
I think this is one step toward addressing @esseivaju 's problem with https://github.com/celeritas-project/celeritas/pull/765#discussion_r1198371330 . I didn't realize till actually opening this MR that it's not quite what he needs...

This could also be the template for the begin/end "batch statistics" work with @helen-brooks .

### CODE REVIEW

File: app/demo-loop/Runner.cc [hunk 1]
```diff
@@ -406,13 +406,10 @@ void Runner::build_diagnostics(RunnerInput const& inp)
...
         auto action_diagnostic = std::make_shared<ActionDiagnostic>(
-            core_params_->action_reg()->next_id(),
-            core_params_->action_reg(),
-            core_params_->particle(),
-            core_params_->max_streams());
+            core_params_->action_reg()->next_id());
 
         // Add to action registry
-        core_params_->action_reg()->insert(action_diagnostic);
+        core_params_->action_reg()->insert_mutable(action_diagnostic);
```
Comment (sethrj) [inline app/demo-loop/Runner.cc]: I really dislike this, but `shared_ptr<ActionInterface const>` and `shared_ptr<ActionInterface>` become ambiguous overloads when passing `shared_ptr<DerivedAction>` or `shared_ptr<DerivedAction const>`.

Comment (esseivaju) [reply app/demo-loop/Runner.cc]: Do you dislike this more?
```c++
    // Register a mutable action
    template<class SPMut,
             class Derived = typename SPMut::element_type,
             class Base = typename SPAction::element_type,
             std::enable_if_t<!std::is_const_v<Derived>
                                  && std::is_base_of_v<Base, Derived>,
                              bool>
             = true>
    void insert(SPMut action)
    {
        CELER_EXPECT(action);
        if (dynamic_cast<BeginRunActionInterface*>(action.get()))
        {
            mutable_actions_.push_back(action);
        }
        return this->insert_impl(std::move(action));
    }

    // Register a const action
    template<class SPConst,
             class Derived = typename SPConst::element_type,
             class Base = typename SPConstAction::element_type,
             std::enable_if_t<std::is_const_v<Derived>
                                  && std::is_base_of_v<Base, Derived>,
                              bool>
             = true>
    void insert(SPConst action)
    {
        CELER_EXPECT(action);
        CELER_VALIDATE(
            !dynamic_cast<BeginRunActionInterface const*>(action.get()),
            << "begin-run action '" << action->label()
            << "' (ID=" << action->action_id().unchecked_get()
            << ") cannot be registered as const");
        return this->insert_impl(std::move(action));
    }
```

Comment (sethrj) [reply app/demo-loop/Runner.cc]: Actually that's great! I think we can make it more readable with a helper traits class. Thanks!

**Review by esseivaju — APPROVED** [review]
Looks good to me. When you say it's not exactly what's needed to correctly size the collection to the number of actions, how so? It seems to me that this would allow to do it.

File: app/demo-loop/Runner.cc (hunk 1 above)
Comment (esseivaju) [reply app/demo-loop/Runner.cc]: Concepts would probably make it even more readable whenever we can use c++ 20.

**Comment by sethrj** [comment]
I wasn't thinking when I said that: since we have an action that manages the sorting, it should work great.

**Review by amandalund — APPROVED** [review]
Nice!  Looks good to me too @sethrj.

**Pull request merged**